src/hotspot/share/tsan/tsanOopMapTable.cpp

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2024, Google and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "tsan/tsanExternalDecls.hpp"
+#include "tsan/tsanOopMap.hpp"
+#include "tsan/tsanOopMapTable.hpp"
+
+TsanOopMapTableKey::TsanOopMapTableKey(oop obj) {
+  _wh = WeakHandle(TsanOopMap::oop_storage(), obj);
+  _obj = obj;
+}
+
+TsanOopMapTableKey::TsanOopMapTableKey(const TsanOopMapTableKey& src) {
+  _wh = src._wh;
+  _obj = src._obj;
+}
+
+void TsanOopMapTableKey::release_weak_handle() const {
+  _wh.release(TsanOopMap::oop_storage());
+}
+
+oop TsanOopMapTableKey::object_no_keepalive() const {
+  return _wh.peek();
+}
+
+void TsanOopMapTableKey::update_obj() {
+  oop obj = _wh.peek();
+  if (obj != nullptr && obj != _obj) {
+    _obj = obj;
+  }
+}
+
+TsanOopMapTable::TsanOopMapTable() : _table(512, 0x3fffffff) {}
+
+void TsanOopMapTable::clear() {
+  struct RemoveAll {
+    bool do_entry(const TsanOopMapTableKey & entry, size_t size) {
+      entry.release_weak_handle();
+      return true;
+    }
+  } remove_all;
+
+  _table.unlink(&remove_all);
+  assert(_table.number_of_entries() == 0, "invariant");
+}
+
+TsanOopMapTable::~TsanOopMapTable() {
+  clear();
+}
+
+bool TsanOopMapTable::add_oop_with_size(oop obj, size_t size) {
+  TsanOopMapTableKey new_entry(obj);
+  bool added;
+  if (obj->fast_no_hash_check()) {
+    added = _table.put_when_absent(new_entry, size);
+  } else {
+    size_t* v = _table.put_if_absent(new_entry, size, &added);
+    *v = size;
+  }
+
+  if (added) {
+    if (_table.maybe_grow(true /* use_large_table_sizes */)) {
+      log_info(tsan)("TsanOopMapTable resize to %d, %d entries",
```

Comment:
Could it be `log_debug(tsan)` as well?

Reply:
Let's keep `log_info`. Resizing occurs not frequently, there's no concern with too much resizing log output. The resizing info maybe useful for finding more adequate table size.